src/app/libp2p_helper/src/bitswap_storage.go

```diff
@@ -0,0 +1,142 @@
+package codanet
+
+import (
+	"fmt"
+
+	lmdbbs "github.com/georgeee/go-bs-lmdb"
+	"github.com/ipfs/go-cid"
+	"github.com/multiformats/go-multihash"
+)
+
+type RootBlockStatus int
+
+const (
+	Partial RootBlockStatus = iota
+	Full
+	Deleting
+)
+
+type BitswapStorage interface {
+	GetStatus(key [32]byte) (RootBlockStatus, error)
+	SetStatus(key [32]byte, value RootBlockStatus) error
+	DeleteStatus(key [32]byte) error
+	DeleteBlocks(keys [][32]byte) error
+	ViewBlock(key [32]byte, callback func([]byte) error) error
+}
+
+type BitswapStorageLmdb lmdbbs.Blockstore
+
+func UnmarshalRootBlockStatus(r []byte) (res RootBlockStatus, err error) {
+	err = fmt.Errorf("wrong root block status retrieved: %v", r)
+	if len(r) != 1 {
+		return
+	}
+	res = RootBlockStatus(r[0])
+	if res == Partial || res == Full || res == Deleting {
+		err = nil
+	}
+	return
+}
+
+func statusKey(key [32]byte) []byte {
+	return append([]byte{BS_STATUS_PREFIX}, key[:]...)
+}
+
+func blockKey(key []byte) []byte {
+	return append([]byte{BS_BLOCK_PREFIX}, key...)
+}
+
+func (bs_ *BitswapStorageLmdb) ViewBlock(key [32]byte, callback func([]byte) error) error {
+	bs := (*lmdbbs.Blockstore)(bs_)
+	return bs.View(BlockHashToCid(key), callback)
+}
+
+func (bs_ *BitswapStorageLmdb) GetStatus(key [32]byte) (res RootBlockStatus, err error) {
+	bs := (*lmdbbs.Blockstore)(bs_)
+	r, err := bs.GetData(statusKey(key))
+	if err != nil {
+		return
+	}
+	res, err = UnmarshalRootBlockStatus(r)
+	return
+}
+
+func (bs_ *BitswapStorageLmdb) DeleteStatus(key [32]byte) error {
+	bs := (*lmdbbs.Blockstore)(bs_)
+	return bs.PutData(statusKey(key), func(prevVal []byte, exists bool) (newVal []byte, newExists bool, err error) {
+		prev, err := UnmarshalRootBlockStatus(prevVal)
+		if err != nil {
+			return
+		}
+		newExists = false
+		if exists && prev != Deleting {
+			err = fmt.Errorf("wrong status deletion from %d", prev)
+		}
+		return
+	})
+}
+
+func (bs_ *BitswapStorageLmdb) SetStatus(key [32]byte, newStatus RootBlockStatus) error {
+	bs := (*lmdbbs.Blockstore)(bs_)
+	return bs.PutData(statusKey(key), func(prevVal []byte, exists bool) (newVal []byte, newExists bool, err error) {
+		var prev RootBlockStatus
+		if exists {
+			prev, err = UnmarshalRootBlockStatus(prevVal)
+			if err != nil {
+				return
+			}
+		}
+		newVal = []byte{byte(newStatus)}
+		newExists = true
+		allowed := false
+		allowed = allowed || (newStatus == Partial && (!exists || prev == Partial))
```

Comment:
It might be nice to extract this into a separate function that encapsulates the state machine's transition guards.

Reply:
Alright